swabbie-core/src/test/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSenderTest.kt

```diff
@@ -0,0 +1,167 @@
+  private val applicationEventPublisher = mock<ApplicationEventPublisher>()
+
+  private val workConfiguration1 = WorkConfigurationTestHelper
+    .generateWorkConfiguration(namespace = "ns1", resourceType = "type1")
+  private val workConfiguration2 = WorkConfigurationTestHelper
+    .generateWorkConfiguration(namespace = "ns2", resourceType = "type2")
+
+  private val notificationService = NotificationSender(
+    lockingService = LockingService.NOOP(),
+    notifier = notifier,
+    clock = clock,
+    applicationEventPublisher = applicationEventPublisher,
+    resourceTrackingRepository = resourceRepository,
+    notificationQueue = notificationQueue,
+    registry = NoopRegistry(),
+    workConfigurations = listOf(workConfiguration1, workConfiguration2)
+  )
+
+  private val now = clock.millis()
+
+  @BeforeEach
+  fun setup() {
+    notificationService
+      .onDiscoveryUpCallback(
+        RemoteStatusChangedEvent(StatusChangeEvent(InstanceInfo.InstanceStatus.DOWN, InstanceInfo.InstanceStatus.UP)))
+  }
+
+  @AfterEach
+  fun cleanup() {
+    reset(
+      notifier,
+      resourceRepository,
+      applicationEventPublisher
+    )
+  }
+
+  @Test
+  fun `should not notify if there is nothing in the notification queue`() {
+    whenever(
+      resourceRepository.getMarkedResources()
+    ) doReturn listOf(
+      createMarkedResource(workConfiguration = workConfiguration2, id = "3", owner = "test@netflix.com")
+    )
+
+    verify(notifier, never()).notify(any())
+    verify(applicationEventPublisher, never()).publishEvent(any<OwnerNotifiedEvent>())
+  }
+
+  @Test
+  fun `should notify`() {
+    val owner1 = "test@netflix.com"
+    val resource1 = createMarkedResource(workConfiguration = workConfiguration1, id = "1", owner = owner1)
+    val resource2 = createMarkedResource(workConfiguration = workConfiguration1, id = "2", owner = owner1)
+    val resource3 = createMarkedResource(workConfiguration = workConfiguration2, id = "3", owner = owner1)
+
+    val resources = listOf(resource1, resource2, resource3)
+
+    notificationQueue.add(
+      NotificationSender.NotificationTask(
+        resourceType = workConfiguration1.resourceType,
+        namespace = workConfiguration1.namespace
+      ))
+
+    whenever(
+      resourceRepository.getMarkedResources()
+    ) doReturn resources
+
+    whenever(
+      notifier.notify(any())
+    ) doReturn Notifier.NotificationResult(owner1, Notifier.NotificationType.EMAIL, success = true)
+
+    notificationService.sendNotifications()
+
+    argumentCaptor<Notifier.Envelope>().apply {
+      verify(notifier, times(1)).notify(capture())
+      val passedResources = firstValue.resources.map { it.first }
+      expectThat(passedResources)
+        .isNotEmpty()
+        .isEqualTo(
+          listOf(resource1, resource2)
+        )
+    }
+
+    // Should publish events for resource1 and resource2 since they match the resource type
```

Comment:
Maybe this test would be better if it was split out into two tests? Like separately test that notification info is populated correctly, and if type doesn't match then the resource is not notified?

Reply:
There are still a couple of scenarios I'd like tested in this class. I will add separate tests.